crates/oxc_minifier/src/plugins/inject_global_variables.rs

```diff
@@ -0,0 +1,229 @@
+    replace_value: Option<CompactStr>,
+}
+
+impl InjectImport {
+    pub fn named_specifier(source: &str, imported: Option<&str>, local: &str) -> InjectImport {
+        InjectImport {
+            source: CompactStr::from(source),
+            specifier: InjectImportSpecifier::Specifier {
+                imported: imported.map(CompactStr::from),
+                local: CompactStr::from(local),
+            },
+            replace_value: Self::replace_name(local),
+        }
+    }
+
+    pub fn namespace_specifier(source: &str, local: &str) -> InjectImport {
+        InjectImport {
+            source: CompactStr::from(source),
+            specifier: InjectImportSpecifier::NamespaceSpecifier { local: CompactStr::from(local) },
+            replace_value: Self::replace_name(local),
+        }
+    }
+
+    pub fn default_specifier(source: &str, local: &str) -> InjectImport {
+        InjectImport {
+            source: CompactStr::from(source),
+            specifier: InjectImportSpecifier::DefaultSpecifier { local: CompactStr::from(local) },
+            replace_value: Self::replace_name(local),
+        }
+    }
+
+    fn replace_name(local: &str) -> Option<CompactStr> {
+        local
+            .contains('.')
+            .then(|| CompactStr::from(format!("$inject_{}", local.replace('.', "_"))))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum InjectImportSpecifier {
+    /// `import { local } from "source"`
+    /// `import { default as local } from "source"` when `imported` is `None`
+    Specifier { imported: Option<CompactStr>, local: CompactStr },
+    /// import * as local from "source"
+    NamespaceSpecifier { local: CompactStr },
+    /// import local from "source"
+    DefaultSpecifier { local: CompactStr },
+}
+
+impl InjectImportSpecifier {
+    fn local(&self) -> &CompactStr {
+        match self {
+            Self::Specifier { local, .. }
+            | Self::NamespaceSpecifier { local, .. }
+            | Self::DefaultSpecifier { local, .. } => local,
+        }
+    }
+}
+
+impl From<&InjectImport> for DotDefine {
+    fn from(inject: &InjectImport) -> Self {
+        let parts = inject.specifier.local().split('.').map(CompactStr::from).collect::<Vec<_>>();
+        let value = inject.replace_value.clone().unwrap();
+        Self { parts, value }
+    }
+}
+
+/// Injects import statements for global variables.
+///
+/// References:
+///
+/// * <https://www.npmjs.com/package/@rollup/plugin-inject>
+pub struct InjectGlobalVariables<'a> {
+    ast: AstBuilder<'a>,
+    config: InjectGlobalVariablesConfig,
+
+    // states
+    /// Dot defines derived from the config.
+    dot_defines: Vec<DotDefine>,
+
+    /// Identifiers for which dot define replaced a member expression.
+    replaced_dot_defines:
+        Vec<(/* identifier of member expression */ CompactStr, /* local */ CompactStr)>,
+}
+
+impl<'a> VisitMut<'a> for InjectGlobalVariables<'a> {
```

Comment:
Because we don't have a mechanism to pass through symbols and scopes in a set of plugins.

We need to properly architect this thing https://github.com/rolldown/rolldown/blob/main/crates/rolldown/src/utils/pre_process_ecma_ast.rs